IllegalStateException when last write fails

**Jetty version(s)**
12.0.6

**Description**
This is almost the same as in #11095 but this time the call stack is different. But while it was this call stack there

java.lang.IllegalStateException: s=HANDLING rs=COMPLETED os=ABORTED is=IDLE awp=false se=false i=false al=0
at org.eclipse.jetty.ee10.servlet.ServletChannelState.completed(ServletChannelState.java:1052) ~[jetty-ee10-servlet-12.0.5.jar:12.0.5]
at org.eclipse.jetty.ee10.servlet.ServletChannel.lambda$handle$3(ServletChannel.java:553) ~[jetty-ee10-servlet-12.0.5.jar:12.0.5]
**at org.eclipse.jetty.util.Callback$3.succeeded(Callback.java:162) ~[jetty-util-12.0.5.jar:12.0.5]
at org.eclipse.jetty.ee10.servlet.HttpOutput.complete(HttpOutput.java:424) ~[jetty-ee10-servlet-12.0.5.jar:12.0.5]**
...

I got this now with a different call stack. Probably because the requests were cancelled by the browser

java.lang.IllegalStateException: s=HANDLING rs=COMPLETED os=ABORTED is=IDLE awp=false se=false i=false al=0
	at org.eclipse.jetty.ee10.servlet.ServletChannelState.completed(ServletChannelState.java:1052)
	**at org.eclipse.jetty.util.Callback$3.failed(Callback.java:168)
	at org.eclipse.jetty.ee10.servlet.HttpOutput.onWriteComplete(HttpOutput.java:252)
	at org.eclipse.jetty.ee10.servlet.HttpOutput$WriteCompleteCB.failed(HttpOutput.java:1760)
	at org.eclipse.jetty.ee10.servlet.HttpOutput$CompleteWriteCompleteCB.failed(HttpOutput.java:1777)
	at org.eclipse.jetty.server.handler.ContextHandler$ScopedContext.accept(ContextHandler.java:1179)
	at org.eclipse.jetty.server.handler.ContextResponse$1.failed(ContextResponse.java:47)
	at org.eclipse.jetty.util.IteratingNestedCallback.onCompleteFailure(IteratingNestedCallback.java:58)**
	at org.eclipse.jetty.server.handler.gzip.GzipResponseAndCallback$GzipBufferCB.onCompleteFailure(GzipResponseAndCallback.java:332)
	at org.eclipse.jetty.util.IteratingCallback.processing(IteratingCallback.java:335)
	at org.eclipse.jetty.util.IteratingCallback.iterate(IteratingCallback.java:231)
	at org.eclipse.jetty.server.handler.gzip.GzipResponseAndCallback.gzip(GzipResponseAndCallback.java:157)
	at org.eclipse.jetty.server.handler.gzip.GzipResponseAndCallback.write(GzipResponseAndCallback.java:135)
	at org.eclipse.jetty.server.Response$Wrapper.write(Response.java:754)
	at org.eclipse.jetty.server.handler.ContextResponse.write(ContextResponse.java:56)
	at org.eclipse.jetty.ee10.servlet.ServletContextResponse.write(ServletContextResponse.java:288)
	at org.eclipse.jetty.ee10.servlet.HttpOutput.channelWrite(HttpOutput.java:204)
	at org.eclipse.jetty.ee10.servlet.HttpOutput.complete(HttpOutput.java:435)
	at org.eclipse.jetty.ee10.servlet.ServletContextResponse.completeOutput(ServletContextResponse.java:212)
	at org.eclipse.jetty.ee10.servlet.ServletChannel.handle(ServletChannel.java:562)
	at org.eclipse.jetty.ee10.servlet.ServletHandler.handle(ServletHandler.java:464)
	at org.eclipse.jetty.security.SecurityHandler.handle(SecurityHandler.java:571)
	at org.eclipse.jetty.ee10.servlet.SessionHandler.handle(SessionHandler.java:703)
	at org.eclipse.jetty.server.handler.ContextHandler.handle(ContextHandler.java:763)
	at org.eclipse.jetty.server.handler.gzip.GzipHandler.handle(GzipHandler.java:597)
	at org.eclipse.jetty.server.handler.ContextHandlerCollection.handle(ContextHandlerCollection.java:181)
	at org.eclipse.jetty.server.Server.handle(Server.java:179)
	at org.eclipse.jetty.server.internal.HttpChannelState$HandlerInvoker.run(HttpChannelState.java:619)
	at org.eclipse.jetty.server.internal.HttpConnection.onFillable(HttpConnection.java:410)
	at org.eclipse.jetty.io.AbstractConnection$ReadCallback.succeeded(AbstractConnection.java:322)
	at org.eclipse.jetty.io.FillInterest.fillable(FillInterest.java:99)
	at org.eclipse.jetty.io.SelectableChannelEndPoint$1.run(SelectableChannelEndPoint.java:53)
	at org.eclipse.jetty.util.thread.strategy.AdaptiveExecutionStrategy.runTask(AdaptiveExecutionStrategy.java:478)
	at org.eclipse.jetty.util.thread.strategy.AdaptiveExecutionStrategy.consumeTask(AdaptiveExecutionStrategy.java:441)
	at org.eclipse.jetty.util.thread.strategy.AdaptiveExecutionStrategy.tryProduce(AdaptiveExecutionStrategy.java:293)
	at org.eclipse.jetty.util.thread.strategy.AdaptiveExecutionStrategy.run(AdaptiveExecutionStrategy.java:201)
	at org.eclipse.jetty.util.thread.ReservedThreadExecutor$ReservedThread.run(ReservedThreadExecutor.java:410)
	at org.eclipse.jetty.util.thread.QueuedThreadPool.runJob(QueuedThreadPool.java:971)
	at org.eclipse.jetty.util.thread.QueuedThreadPool$Runner.doRunJob(QueuedThreadPool.java:1201)
	at org.eclipse.jetty.util.thread.QueuedThreadPool$Runner.run(QueuedThreadPool.java:1156)
	at java.base/java.lang.Thread.run(Thread.java:833)


I haven't tried to verify whether it was really caused by the GZip filter too in this case yet as that is not as easy to reproduce

Actually there was a mix of static and dynamicl content in the previous example. The theme.css I mention is actually a user configurable CSS and served from a servlet filter. But other CSS, JS etc is really static content and served by the Jetty DefaultServlet. The first time I saw this error in the logs it was for such static content but when I tried to reproduce it was reported only for the dynamic content. Don't ask me why. But just to make sure i switched back to the default theme which is a static resource and run then brute force test again. And it is as easy to reproduce as before with the static resources.

What i noticed is that when it happens during the brute force test it is reported only for a single file

```
2024/02/03 09:10:54.320 [33mWARN  [ServletChannel                ]  /gatekeeper/ui/gatekeeper/js/jquery.min.js
java.lang.IllegalStateException: s=HANDLING rs=COMPLETED os=ABORTED is=IDLE awp=false se=false i=false al=0

[m2024/02/03 09:10:54.492 [33mWARN  [ServletChannel                ]  /gatekeeper/ui/gatekeeper/js/jquery.min.js
java.lang.IllegalStateException: s=HANDLING rs=COMPLETED os=ABORTED is=IDLE awp=false se=false i=false al=0

[m2024/02/03 09:10:54.586 [33mWARN  [ServletChannel                ]  /gatekeeper/ui/gatekeeper/js/jquery.min.js
java.lang.IllegalStateException: s=HANDLING rs=COMPLETED os=ABORTED is=IDLE awp=false se=false i=false al=0

[m2024/02/03 09:10:54.649 [33mWARN  [ServletChannel                ]  /gatekeeper/ui/gatekeeper/js/jquery.min.js
java.lang.IllegalStateException: s=HANDLING rs=COMPLETED os=ABORTED is=IDLE awp=false se=false i=false al=0

[m2024/02/03 09:10:54.984 [33mWARN  [ServletChannel                ]  /gatekeeper/ui/gatekeeper/js/jquery.min.js
java.lang.IllegalStateException: s=HANDLING rs=COMPLETED os=ABORTED is=IDLE awp=false se=false i=false al=0

[m2024/02/03 09:10:55.084 [33mWARN  [ServletChannel                ]  /gatekeeper/ui/gatekeeper/js/jquery.min.js
java.lang.IllegalStateException: s=HANDLING rs=COMPLETED os=ABORTED is=IDLE awp=false se=false i=false al=0

[m2024/02/03 09:10:55.152 [33mWARN  [ServletChannel                ]  /gatekeeper/ui/gatekeeper/js/jquery.min.js
java.lang.IllegalStateException: s=HANDLING rs=COMPLETED os=ABORTED is=IDLE awp=false se=false i=false al=0

[m2024/02/03 09:10:55.186 [33mWARN  [ServletChannel                ]  /gatekeeper/ui/gatekeeper/js/jquery.min.js
java.lang.IllegalStateException: s=HANDLING rs=COMPLETED os=ABORTED is=IDLE awp=false se=false i=false al=0

[m2024/02/03 09:10:55.216 [33mWARN  [ServletChannel                ]  /gatekeeper/ui/gatekeeper/js/jquery.min.js
java.lang.IllegalStateException: s=HANDLING rs=COMPLETED os=ABORTED is=IDLE awp=false se=false i=false al=0

[m2024/02/03 09:10:55.245 [33mWARN  [ServletChannel                ]  /gatekeeper/ui/gatekeeper/js/jquery.min.js
java.lang.IllegalStateException: s=HANDLING rs=COMPLETED os=ABORTED is=IDLE awp=false se=false i=false al=0
```

So i changed the order of the references but that didn't change anything. But when I made one of the other files noticeable bigger so that it was even bigger than the jquery.min it started to fail for both files

```
2024/02/03 09:16:56.278 [33mWARN  [ServletChannel                ]  /gatekeeper/ui/gatekeeper/js/util.js
java.lang.IllegalStateException: s=HANDLING rs=COMPLETED os=ABORTED is=IDLE awp=false se=false i=false al=0

[m2024/02/03 09:16:56.281 [33mWARN  [ServletChannel                ]  /gatekeeper/ui/gatekeeper/js/jquery.min.js
java.lang.IllegalStateException: s=HANDLING rs=COMPLETED os=ABORTED is=IDLE awp=false se=false i=false al=0

[m2024/02/03 09:16:56.378 [33mWARN  [ServletChannel                ]  /gatekeeper/ui/gatekeeper/js/util.js
java.lang.IllegalStateException: s=HANDLING rs=COMPLETED os=ABORTED is=IDLE awp=false se=false i=false al=0

[m2024/02/03 09:16:56.381 [33mWARN  [ServletChannel                ]  /gatekeeper/ui/gatekeeper/js/jquery.min.js
java.lang.IllegalStateException: s=HANDLING rs=COMPLETED os=ABORTED is=IDLE awp=false se=false i=false al=0

[m2024/02/03 09:16:57.073 [33mWARN  [ServletChannel                ]  /gatekeeper/ui/gatekeeper/js/util.js
java.lang.IllegalStateException: s=HANDLING rs=COMPLETED os=ABORTED is=IDLE awp=false se=false i=false al=0

[m2024/02/03 09:16:57.363 [33mWARN  [ServletChannel                ]  /gatekeeper/ui/gatekeeper/js/util.js
java.lang.IllegalStateException: s=HANDLING rs=COMPLETED os=ABORTED is=IDLE awp=false se=false i=false al=0
```

So I started playing with the file size to see whether that really means something. And guess what: up to 32.767 bytes it seems to be fine. But once it reaches 32.768 or more the error starts to appear. I also made a test by referencing two versions of the file in the JSP, one 32.767 bytes and one 32.768 and only the 32.768 version fails

```
2024/02/03 09:43:18.032 [33mWARN  [ServletChannel                ]  /gatekeeper/ui/gatekeeper/js/util-32768.js
java.lang.IllegalStateException: s=HANDLING rs=COMPLETED os=ABORTED is=IDLE awp=false se=false i=false al=0

[m2024/02/03 09:43:18.097 [33mWARN  [ServletChannel                ]  /gatekeeper/ui/gatekeeper/js/util-32768.js
java.lang.IllegalStateException: s=HANDLING rs=COMPLETED os=ABORTED is=IDLE awp=false se=false i=false al=0

[m2024/02/03 09:43:18.623 [33mWARN  [ServletChannel                ]  /gatekeeper/ui/gatekeeper/js/util-32768.js
java.lang.IllegalStateException: s=HANDLING rs=COMPLETED os=ABORTED is=IDLE awp=false se=false i=false al=0

[m2024/02/03 09:43:18.853 [33mWARN  [ServletChannel                ]  /gatekeeper/ui/gatekeeper/js/util-32768.js
java.lang.IllegalStateException: s=HANDLING rs=COMPLETED os=ABORTED is=IDLE awp=false se=false i=false al=0

[m2024/02/03 09:43:19.150 [33mWARN  [ServletChannel                ]  /gatekeeper/ui/gatekeeper/js/util-32768.js
java.lang.IllegalStateException: s=HANDLING rs=COMPLETED os=ABORTED is=IDLE awp=false se=false i=false al=0
```

[jetty.log](https://github.com/jetty/jetty.project/files/14150801/jetty.log)

```
2024/02/03 09:47:31.875  DEBUG [] [http-server-204] org.eclipse.jetty.ee10.servlet.HttpOutput write(array HeapByteBuffer@5e12b804[p=0,l=4096,c=4096,r=4096]={<<<  }\n\t\t\t  }\n\t\t\t});\n\t}*/\n}\n...\n\tregisterCapsLockWarni>>>})
2024/02/03 09:47:31.875  DEBUG [] [http-server-204] org.eclipse.jetty.ee10.servlet.HttpOutput write(array) s=OPEN,api=BLOCKING,sc=false,e=null aggregated !flush Buffer@41e8b4aa[rc=1,DirectByteBuffer@7f7f7824[p=0,l=28672,c=32768,r=28672]={<<<function registerCapsLock...\n\tregisterCapsLockWarni>>>ng();\n\tfi...gskj111}]
2024/02/03 09:47:31.875  DEBUG [] [http-server-204] org.eclipse.jetty.ee10.servlet.HttpOutput write(array HeapByteBuffer@1d43bfb[p=0,l=4096,c=4096,r=4096]={<<<ng();\n\tfixPageDivHeight()...}\n\nsdfsdfasdkjfhgskj111>>>})
2024/02/03 09:47:31.875  DEBUG [] [http-server-204] org.eclipse.jetty.ee10.servlet.HttpOutput write(array) s=OPEN,api=BLOCKED,sc=false,e=null last=false agg=true flush=true async=false, len=0 Buffer@41e8b4aa[rc=1,DirectByteBuffer@7f7f7824[p=0,l=32768,c=32768,r=32768]={<<<function registerCapsLock...}\n\nsdfsdfasdkjfhgskj111>>>}]
2024/02/03 09:47:31.875  DEBUG [] [http-server-197] org.eclipse.jetty.ee10.servlet.HttpOutput onWriteComplete(true,null) s=CLOSING,api=BLOCKED,sc=false,e=null->s=CLOSED,api=BLOCKING,sc=false,e=null c=null cb=Callback@253a046e{NON_BLOCKING, org.eclipse.jetty.ee10.servlet.ServletChannel$$Lambda$1192/0x00000008018e5000@600616b,org.eclipse.jetty.ee10.servlet.ServletChannel$$Lambda$1197/0x00000008018e5aa0@2bfe2958} w=false
2024/02/03 09:47:31.875  DEBUG [] [http-server-190] org.eclipse.jetty.ee10.servlet.HttpOutput onWriteComplete(true,null) s=CLOSING,api=BLOCKED,sc=false,e=null->s=CLOSED,api=BLOCKING,sc=false,e=null c=null cb=Callback@3374dc06{NON_BLOCKING, org.eclipse.jetty.ee10.servlet.ServletChannel$$Lambda$1192/0x00000008018e5000@784fab59,org.eclipse.jetty.ee10.servlet.ServletChannel$$Lambda$1197/0x00000008018e5aa0@1fb16a45} w=false
2024/02/03 09:47:31.876  DEBUG [] [http-server-197] org.eclipse.jetty.ee10.servlet.ServletChannelState completed ServletChannelState@57ca552c{s=HANDLING rs=COMPLETING os=OPEN is=IDLE awp=false se=false i=false al=0}
2024/02/03 09:47:31.876  DEBUG [] [http-server-190] org.eclipse.jetty.ee10.servlet.ServletChannelState completed ServletChannelState@2a4b01d0{s=HANDLING rs=COMPLETING os=OPEN is=IDLE awp=false se=false i=false al=0}
2024/02/03 09:47:31.876  DEBUG [] [http-server-197] org.eclipse.jetty.ee10.servlet.ServletChannelState unhandle ServletChannelState@57ca552c{s=HANDLING rs=COMPLETED os=OPEN is=IDLE awp=false se=false i=false al=0}
2024/02/03 09:47:31.876  DEBUG [] [http-server-197] org.eclipse.jetty.ee10.servlet.ServletChannelState nextAction(false) TERMINATED ServletChannelState@57ca552c{s=IDLE rs=COMPLETED os=OPEN is=IDLE awp=false se=false i=false al=0}
2024/02/03 09:47:31.876  DEBUG [] [http-server-190] org.eclipse.jetty.ee10.servlet.ServletChannelState unhandle ServletChannelState@2a4b01d0{s=HANDLING rs=COMPLETED os=OPEN is=IDLE awp=false se=false i=false al=0}
2024/02/03 09:47:31.876  DEBUG [] [http-server-197] org.eclipse.jetty.ee10.servlet.ServletChannel action TERMINATED ServletChannel@10a29ec{s=ServletChannelState@57ca552c{s=IDLE rs=COMPLETED os=OPEN is=IDLE awp=false se=false i=false al=0},r=0,c=true/false,a=IDLE,uri=http://localhost:8800/gatekeeper/ui/gatekeeper/js/util-32767.js?1706949786861,age=1}
2024/02/03 09:47:31.876  DEBUG [] [http-server-197] org.eclipse.jetty.ee10.servlet.ServletChannel onCompleted for /gatekeeper/ui/gatekeeper/js/util-32767.js written app=32767 net=2166
2024/02/03 09:47:31.876  DEBUG [] [http-server-190] org.eclipse.jetty.ee10.servlet.ServletChannelState nextAction(false) TERMINATED ServletChannelState@2a4b01d0{s=IDLE rs=COMPLETED os=OPEN is=IDLE awp=false se=false i=false al=0}
2024/02/03 09:47:31.876  DEBUG [] [http-server-190] org.eclipse.jetty.ee10.servlet.ServletChannel action TERMINATED ServletChannel@734e798f{s=ServletChannelState@2a4b01d0{s=IDLE rs=COMPLETED os=OPEN is=IDLE awp=false se=false i=false al=0},r=0,c=true/false,a=IDLE,uri=http://localhost:8800/gatekeeper/ui/gatekeeper/themes/default/theme.css?1706949786861,age=1}
2024/02/03 09:47:31.876  DEBUG [] [http-server-190] org.eclipse.jetty.ee10.servlet.ServletChannel onCompleted for /gatekeeper/ui/gatekeeper/themes/default/theme.css written app=11647 net=2549
2024/02/03 09:47:31.876  DEBUG [] [http-server-204] org.eclipse.jetty.ee10.servlet.HttpOutput onWriteComplete(false,null) s=OPEN,api=BLOCKED,sc=false,e=null->s=OPEN,api=BLOCKING,sc=false,e=null c=null cb=null w=false
2024/02/03 09:47:31.876  DEBUG [] [http-server-204] org.eclipse.jetty.ee10.servlet.ServletChannelState unhandle ServletChannelState@1f2677b2{s=HANDLING rs=BLOCKING os=OPEN is=IDLE awp=false se=false i=true al=0}
2024/02/03 09:47:31.876  DEBUG [] [http-server-204] org.eclipse.jetty.ee10.servlet.ServletChannelState nextAction(false) COMPLETE ServletChannelState@1f2677b2{s=HANDLING rs=COMPLETING os=OPEN is=IDLE awp=false se=false i=false al=0}
2024/02/03 09:47:31.876  DEBUG [] [http-server-204] org.eclipse.jetty.ee10.servlet.ServletChannel action COMPLETE ServletChannel@5a17429e{s=ServletChannelState@1f2677b2{s=HANDLING rs=COMPLETING os=OPEN is=IDLE awp=false se=false i=false al=0},r=0,c=false/false,a=HANDLING,uri=http://localhost:8800/gatekeeper/ui/gatekeeper/js/util-32768.js?1706949786861,age=1}
2024/02/03 09:47:31.876  DEBUG [] [http-server-204] org.eclipse.jetty.ee10.servlet.HttpOutput complete(Callback@450792b{NON_BLOCKING, org.eclipse.jetty.ee10.servlet.ServletChannel$$Lambda$1192/0x00000008018e5000@7caa7d15,org.eclipse.jetty.ee10.servlet.ServletChannel$$Lambda$1197/0x00000008018e5aa0@37b3281c}) s=CLOSING,api=BLOCKED,sc=false,e=null s=false e=null, c=HeapByteBufferR@1442f436[p=0,l=0,c=0,r=0]={<<<>>>}
2024/02/03 09:47:31.876  DEBUG [] [http-server-197] org.eclipse.jetty.ee10.servlet.ServletChannel !handle TERMINATED ServletChannel@10a29ec{s=ServletChannelState@57ca552c{s=IDLE rs=COMPLETED os=COMPLETED is=IDLE awp=false se=false i=false al=0},r=0,c=true/true,a=IDLE,uri=http://localhost:8800/gatekeeper/ui/gatekeeper/js/util-32767.js?1706949786861,age=2}
2024/02/03 09:47:31.876  DEBUG [] [http-server-204] org.eclipse.jetty.ee10.servlet.ServletChannelState abort=true ServletChannelState@1f2677b2{s=HANDLING rs=COMPLETING os=ABORTED is=IDLE awp=false se=false i=false al=0}
org.eclipse.jetty.io.EofException: null
	at org.eclipse.jetty.io.SocketChannelEndPoint.flush(SocketChannelEndPoint.java:117) ~[?:?]
	at org.eclipse.jetty.io.WriteFlusher.flush(WriteFlusher.java:422) ~[?:?]
	at org.eclipse.jetty.io.WriteFlusher.write(WriteFlusher.java:275) ~[?:?]
	at org.eclipse.jetty.io.WriteFlusher.write(WriteFlusher.java:254) ~[?:?]
	...
Caused by: java.io.IOException: An established connection was aborted by the software in your host machine
	at java.base/sun.nio.ch.SocketDispatcher.writev0(Native Method) ~[?:?]
	at java.base/sun.nio.ch.SocketDispatcher.writev(SocketDispatcher.java:58) ~[?:?]
	at java.base/sun.nio.ch.IOUtil.write(IOUtil.java:217) ~[?:?]
	at java.base/sun.nio.ch.IOUtil.write(IOUtil.java:153) ~[?:?]
	at java.base/sun.nio.ch.SocketChannelImpl.write(SocketChannelImpl.java:563) ~[?:?]
	at java.base/java.nio.channels.SocketChannel.write(SocketChannel.java:642) ~[?:?]
	at org.eclipse.jetty.io.SocketChannelEndPoint.flush(SocketChannelEndPoint.java:111) ~[?:?]
	... 50 more
2024/02/03 09:47:31.876  DEBUG [] [http-server-197] org.eclipse.jetty.ee10.servlet.ServletChannelState recycle ServletChannelState@57ca552c{s=IDLE rs=COMPLETED os=COMPLETED is=IDLE awp=false se=false i=false al=0}
2024/02/03 09:47:31.876  DEBUG [] [http-server-197] org.eclipse.jetty.ee10.servlet.HttpInput recycle HttpInput@1822207510 cs=ServletChannelState@57ca552c{s=IDLE rs=BLOCKING os=IDLE is=IDLE awp=false se=false i=true al=0} cp=org.eclipse.jetty.ee10.servlet.BlockingContentProducer@1b7aaf58 eof=false
2024/02/03 09:47:31.876  DEBUG [] [http-server-197] org.eclipse.jetty.ee10.servlet.BlockingContentProducer recycling org.eclipse.jetty.ee10.servlet.BlockingContentProducer@1b7aaf58
2024/02/03 09:47:31.876  DEBUG [] [http-server-197] org.eclipse.jetty.ee10.servlet.AsyncContentProducer recycling AsyncContentProducer@8efcaf9[c=null]
2024/02/03 09:47:31.876  DEBUG [] [http-server-204] org.eclipse.jetty.ee10.servlet.HttpOutput onWriteComplete(true,org.eclipse.jetty.io.EofException) s=CLOSING,api=BLOCKED,sc=false,e=null->s=CLOSED,api=BLOCKING,sc=false,e=null c=null cb=Callback@450792b{NON_BLOCKING, org.eclipse.jetty.ee10.servlet.ServletChannel$$Lambda$1192/0x00000008018e5000@7caa7d15,org.eclipse.jetty.ee10.servlet.ServletChannel$$Lambda$1197/0x00000008018e5aa0@37b3281c} w=false
org.eclipse.jetty.io.EofException: null
	at org.eclipse.jetty.io.SocketChannelEndPoint.flush(SocketChannelEndPoint.java:117) ~[?:?]
	at org.eclipse.jetty.io.WriteFlusher.flush(WriteFlusher.java:422) ~[?:?]
	at org.eclipse.jetty.io.WriteFlusher.write(WriteFlusher.java:275) ~[?:?]
	at org.eclipse.jetty.io.WriteFlusher.write(WriteFlusher.java:254) ~[?:?]
	at org.eclipse.jetty.io.AbstractEndPoint.write(AbstractEndPoint.java:368) ~[?:?]
	...
Caused by: java.io.IOException: An established connection was aborted by the software in your host machine
	at java.base/sun.nio.ch.SocketDispatcher.writev0(Native Method) ~[?:?]
	at java.base/sun.nio.ch.SocketDispatcher.writev(SocketDispatcher.java:58) ~[?:?]
	at java.base/sun.nio.ch.IOUtil.write(IOUtil.java:217) ~[?:?]
	at java.base/sun.nio.ch.IOUtil.write(IOUtil.java:153) ~[?:?]
	at java.base/sun.nio.ch.SocketChannelImpl.write(SocketChannelImpl.java:563) ~[?:?]
	at java.base/java.nio.channels.SocketChannel.write(SocketChannel.java:642) ~[?:?]
	at org.eclipse.jetty.io.SocketChannelEndPoint.flush(SocketChannelEndPoint.java:111) ~[?:?]
	... 50 more
2024/02/03 09:47:31.876  DEBUG [] [http-server-204] org.eclipse.jetty.ee10.servlet.ServletChannelState completed ServletChannelState@1f2677b2{s=HANDLING rs=COMPLETED os=ABORTED is=IDLE awp=false se=false i=false al=0}
2024/02/03 09:47:31.876  WARN  [] [http-server-204] org.eclipse.jetty.ee10.servlet.ServletChannel /gatekeeper/ui/gatekeeper/js/util-32768.js
java.lang.IllegalStateException: s=HANDLING rs=COMPLETED os=ABORTED is=IDLE awp=false se=false i=false al=0
	at org.eclipse.jetty.ee10.servlet.ServletChannelState.completed(ServletChannelState.java:1052) ~[?:?]
	at org.eclipse.jetty.util.Callback$3.failed(Callback.java:168) ~[?:?]
	at org.eclipse.jetty.ee10.servlet.HttpOutput.onWriteComplete(HttpOutput.java:252) ~[?:?]
	at org.eclipse.jetty.ee10.servlet.HttpOutput$WriteCompleteCB.failed(HttpOutput.java:1760) ~[?:?]
	at org.eclipse.jetty.ee10.servlet.HttpOutput$CompleteWriteCompleteCB.failed(HttpOutput.java:1777) ~[?:?]
	at org.eclipse.jetty.server.handler.ContextHandler$ScopedContext.accept(ContextHandler.java:1179) ~[?:?]
	at org.eclipse.jetty.server.handler.ContextResponse$1.failed(ContextResponse.java:47) ~[?:?]
	at org.eclipse.jetty.util.IteratingNestedCallback.onCompleteFailure(IteratingNestedCallback.java:58) ~[?:?]
	at org.eclipse.jetty.server.handler.gzip.GzipResponseAndCallback$GzipBufferCB.onCompleteFailure(GzipResponseAndCallback.java:332) ~[?:?]
	at org.eclipse.jetty.util.IteratingCallback.processing(IteratingCallback.java:335) ~[?:?]
	at org.eclipse.jetty.util.IteratingCallback.iterate(IteratingCallback.java:231) ~[?:?]
	at org.eclipse.jetty.server.handler.gzip.GzipResponseAndCallback.gzip(GzipResponseAndCallback.java:157) ~[?:?]
	at org.eclipse.jetty.server.handler.gzip.GzipResponseAndCallback.write(GzipResponseAndCallback.java:135) ~[?:?]
	...
2024/02/03 09:47:31.876  DEBUG [] [http-server-190] org.eclipse.jetty.ee10.servlet.ServletChannel !handle TERMINATED ServletChannel@734e798f{s=ServletChannelState@2a4b01d0{s=IDLE rs=COMPLETED os=COMPLETED is=IDLE awp=false se=false i=false al=0},r=0,c=true/true,a=IDLE,uri=http://localhost:8800/gatekeeper/ui/gatekeeper/themes/default/theme.css?1706949786861,age=2}
2024/02/03 09:47:31.876  DEBUG [] [http-server-204] org.eclipse.jetty.ee10.servlet.ServletChannelState abort=false ServletChannelState@1f2677b2{s=HANDLING rs=COMPLETED os=ABORTED is=IDLE awp=false se=false i=false al=0}
java.lang.IllegalStateException: s=HANDLING rs=COMPLETED os=ABORTED is=IDLE awp=false se=false i=false al=0
	at org.eclipse.jetty.ee10.servlet.ServletChannelState.completed(ServletChannelState.java:1052) ~[?:?]
	at org.eclipse.jetty.util.Callback$3.failed(Callback.java:168) ~[?:?]
	at org.eclipse.jetty.ee10.servlet.HttpOutput.onWriteComplete(HttpOutput.java:252) ~[?:?]
	at org.eclipse.jetty.ee10.servlet.HttpOutput$WriteCompleteCB.failed(HttpOutput.java:1760) ~[?:?]
	at org.eclipse.jetty.ee10.servlet.HttpOutput$CompleteWriteCompleteCB.failed(HttpOutput.java:1777) ~[?:?]
	at org.eclipse.jetty.server.handler.ContextHandler$ScopedContext.accept(ContextHandler.java:1179) ~[?:?]
	at org.eclipse.jetty.server.handler.ContextResponse$1.failed(ContextResponse.java:47) ~[?:?]
	at org.eclipse.jetty.util.IteratingNestedCallback.onCompleteFailure(IteratingNestedCallback.java:58) ~[?:?]
	at org.eclipse.jetty.server.handler.gzip.GzipResponseAndCallback$GzipBufferCB.onCompleteFailure(GzipResponseAndCallback.java:332) ~[?:?]
	at org.eclipse.jetty.util.IteratingCallback.processing(IteratingCallback.java:335) ~[?:?]
	at org.eclipse.jetty.util.IteratingCallback.iterate(IteratingCallback.java:231) ~[?:?]
	at org.eclipse.jetty.server.handler.gzip.GzipResponseAndCallback.gzip(GzipResponseAndCallback.java:157) ~[?:?]
	at org.eclipse.jetty.server.handler.gzip.GzipResponseAndCallback.write(GzipResponseAndCallback.java:135) ~[?:?]
	at org.eclipse.jetty.server.Response$Wrapper.write(Response.java:754) ~[?:?]
	....
2024/02/03 09:47:31.876  DEBUG [] [http-server-204] org.eclipse.jetty.ee10.servlet.ServletChannelState unhandle ServletChannelState@1f2677b2{s=HANDLING rs=COMPLETED os=ABORTED is=IDLE awp=false se=false i=false al=0}
2024/02/03 09:47:31.876  DEBUG [] [http-server-204] org.eclipse.jetty.ee10.servlet.ServletChannelState nextAction(false) TERMINATED ServletChannelState@1f2677b2{s=IDLE rs=COMPLETED os=ABORTED is=IDLE awp=false se=false i=false al=0}
2024/02/03 09:47:31.876  DEBUG [] [http-server-204] org.eclipse.jetty.ee10.servlet.ServletChannel action TERMINATED ServletChannel@5a17429e{s=ServletChannelState@1f2677b2{s=IDLE rs=COMPLETED os=ABORTED is=IDLE awp=false se=false i=false al=0},r=0,c=true/false,a=IDLE,uri=http://localhost:8800/gatekeeper/ui/gatekeeper/js/util-32768.js?1706949786861,age=1}
2024/02/03 09:47:31.876  DEBUG [] [http-server-190] org.eclipse.jetty.ee10.servlet.ServletChannelState recycle ServletChannelState@2a4b01d0{s=IDLE rs=COMPLETED os=COMPLETED is=IDLE awp=false se=false i=false al=0}
2024/02/03 09:47:31.876  DEBUG [] [http-server-204] org.eclipse.jetty.ee10.servlet.ServletChannel onCompleted for /gatekeeper/ui/gatekeeper/js/util-32768.js written app=32768 net=2167
2024/02/03 09:47:31.876  DEBUG [] [http-server-190] org.eclipse.jetty.ee10.servlet.HttpInput recycle HttpInput@1885870698 cs=ServletChannelState@2a4b01d0{s=IDLE rs=BLOCKING os=IDLE is=IDLE awp=false se=false i=true al=0} cp=org.eclipse.jetty.ee10.servlet.BlockingContentProducer@2a4299a4 eof=false
2024/02/03 09:47:31.876  DEBUG [] [http-server-190] org.eclipse.jetty.ee10.servlet.BlockingContentProducer recycling org.eclipse.jetty.ee10.servlet.BlockingContentProducer@2a4299a4
2024/02/03 09:47:31.876  DEBUG [] [http-server-190] org.eclipse.jetty.ee10.servlet.AsyncContentProducer recycling AsyncContentProducer@43c590d7[c=null]
2024/02/03 09:47:31.877  DEBUG [] [http-server-204] org.eclipse.jetty.ee10.servlet.ServletChannel !handle TERMINATED ServletChannel@5a17429e{s=ServletChannelState@1f2677b2{s=IDLE rs=COMPLETED os=ABORTED is=IDLE awp=false se=false i=false al=0},r=0,c=true/false,a=IDLE,uri=http://localhost:8800/gatekeeper/ui/gatekeeper/js/util-32768.js?1706949786861,age=2}
2024/02/03 09:47:31.877  DEBUG [] [http-server-204] org.eclipse.jetty.ee10.servlet.ServletChannelState recycle ServletChannelState@1f2677b2{s=IDLE rs=COMPLETED os=ABORTED is=IDLE awp=false se=false i=false al=0}
2024/02/03 09:47:31.877  DEBUG [] [http-server-204] org.eclipse.jetty.ee10.servlet.HttpInput recycle HttpInput@1559729658 cs=ServletChannelState@1f2677b2{s=IDLE rs=BLOCKING os=ABORTED is=IDLE awp=false se=false i=true al=0} cp=org.eclipse.jetty.ee10.servlet.BlockingContentProducer@2db595f4 eof=false
2024/02/03 09:47:31.877  DEBUG [] [http-server-204] org.eclipse.jetty.ee10.servlet.BlockingContentProducer recycling org.eclipse.jetty.ee10.servlet.BlockingContentProducer@2db595f4
2024/02/03 09:47:31.877  DEBUG [] [http-server-204] org.eclipse.jetty.ee10.servlet.AsyncContentProducer recycling AsyncContentProducer@3aaf5874[c=null]
```


There are is no Expires headers. But in the end that would wouldn't change much about the underlying problem when as the brute force test is using CTRL-F5 to force a complete reload. It doesn't happen with a standard reload even though FF still logs NS_BINDING_ABORTED even on a simple reload with just F5

--- a/jetty-core/jetty-server/src/main/java/org/eclipse/jetty/server/SizeLimitHandler.java
+++ b/jetty-core/jetty-server/src/main/java/org/eclipse/jetty/server/SizeLimitHandler.java
@@ -143,8 +143,8 @@ public class SizeLimitHandler extends Handler.Wrapper
             {
                 if (_responseLimit >= 0 && (_written + content.remaining())  > _responseLimit)
                 {
-                    callback.failed(new HttpException.RuntimeException(HttpStatus.INTERNAL_SERVER_ERROR_500, "Response body is too large: " +
-                        _written + content.remaining() + ">" + _responseLimit));
+                    String message = "Response body is too large: %d > %d".formatted(_written + content.remaining(), _responseLimit);
+                    callback.failed(new HttpException.RuntimeException(HttpStatus.INTERNAL_SERVER_ERROR_500, message));
                     return;
                 }
                 _written += content.remaining();

--- a/jetty-core/jetty-server/src/main/java/org/eclipse/jetty/server/SizeLimitHandler.java
+++ b/jetty-core/jetty-server/src/main/java/org/eclipse/jetty/server/SizeLimitHandler.java
@@ -143,7 +143,7 @@ public class SizeLimitHandler extends Handler.Wrapper
             {
                 if (_responseLimit >= 0 && (_written + content.remaining())  > _responseLimit)
                 {
-                    String message = "Response body is too large: %d > %d".formatted(_written + content.remaining(), _responseLimit);
+                    String message = "Response body is too large: %d>%d".formatted(_written + content.remaining(), _responseLimit);
                     callback.failed(new HttpException.RuntimeException(HttpStatus.INTERNAL_SERVER_ERROR_500, message));
                     return;
                 }

--- a/jetty-core/jetty-server/src/test/java/org/eclipse/jetty/server/handler/SizeLimitHandlerTest.java
+++ b/jetty-core/jetty-server/src/test/java/org/eclipse/jetty/server/handler/SizeLimitHandlerTest.java
@@ -80,7 +80,7 @@ public class SizeLimitHandlerTest
         _contextHandler.setHandler(new Handler.Abstract()
         {
             @Override
-            public boolean handle(Request request, Response response, Callback callback) throws Exception
+            public boolean handle(Request request, Response response, Callback callback)
             {
                 response.write(true, BufferUtil.toBuffer("Hello World"), callback);
                 return true;
@@ -99,7 +99,7 @@ public class SizeLimitHandlerTest
         _contextHandler.setHandler(new Handler.Abstract()
         {
             @Override
-            public boolean handle(Request request, Response response, Callback callback) throws Exception
+            public boolean handle(Request request, Response response, Callback callback)
             {
                 response.getHeaders().put(HttpHeader.CONTENT_LENGTH, 8 * 1024 + 1);
                 fail();
@@ -120,9 +120,8 @@ public class SizeLimitHandlerTest
         _contextHandler.setHandler(new Handler.Abstract()
         {
             @Override
-            public boolean handle(Request request, Response response, Callback callback) throws Exception
+            public boolean handle(Request request, Response response, Callback callback)
             {
-
                 response.write(true, ByteBuffer.wrap(new byte[8 * 1024 + 1]), callback);
                 return true;
             }
@@ -141,7 +140,7 @@ public class SizeLimitHandlerTest
         _contextHandler.setHandler(new Handler.Abstract()
         {
             @Override
-            public boolean handle(Request request, Response response, Callback callback) throws Exception
+            public boolean handle(Request request, Response response, Callback callback)
             {
                 byte[] data = new byte[1024];
                 Arrays.fill(data, (byte)'X');
@@ -199,10 +198,12 @@ public class SizeLimitHandlerTest
         });
         _server.start();
         HttpTester.Response response = HttpTester.parseResponse(
-            _local.getResponse("POST /ctx/hello HTTP/1.0\r\n" +
-                "Content-Length: 8\r\n" +
-                "\r\n" +
-                "123456\r\n"));
+            _local.getResponse("""
+                POST /ctx/hello HTTP/1.0\r
+                Content-Length: 8\r
+                \r
+                123456\r
+                """));
         assertThat(response.getStatus(), equalTo(200));
         assertThat(response.getContent(), containsString("OK 8"));
     }
@@ -223,10 +224,11 @@ public class SizeLimitHandlerTest
         });
         _server.start();
         HttpTester.Response response = HttpTester.parseResponse(
-            _local.getResponse("POST /ctx/hello HTTP/1.0\r\n" +
-                "Content-Length: 32768\r\n" +
-                "\r\n" +
-                "123456..."));
+            _local.getResponse("""
+                POST /ctx/hello HTTP/1.0\r
+                Content-Length: 32768\r
+                \r
+                123456..."""));
         assertThat(response.getStatus(), equalTo(413));
         assertThat(response.getContent(), containsString("32768&gt;8192"));
     }
@@ -247,11 +249,13 @@ public class SizeLimitHandlerTest
         });
         _server.start();
 
-        try (LocalConnector.LocalEndPoint endp = _local.executeRequest(
-            "POST /ctx/hello HTTP/1.1\r\n" +
-                "Host: localhost\r\n" +
-                "Transfer-Encoding: chunked\r\n" +
-                "\r\n"))
+        try (LocalConnector.LocalEndPoint endPoint = _local.executeRequest(
+            """
+                POST /ctx/hello HTTP/1.1\r
+                Host: localhost\r
+                Transfer-Encoding: chunked\r
+                \r
+                """))
         {
             byte[] data = new byte[1024];
             Arrays.fill(data, (byte)'X');
@@ -259,9 +263,9 @@ public class SizeLimitHandlerTest
             String text = new String(data, 0, 1024, Charset.defaultCharset());
 
             for (int i = 0; i < 9; i++)
-                endp.addInput("400\r\n" + text + "\r\n");
+                endPoint.addInput("400\r\n" + text + "\r\n");
 
-            HttpTester.Response response = HttpTester.parseResponse(endp.getResponse());
+            HttpTester.Response response = HttpTester.parseResponse(endPoint.getResponse());
 
             assertThat(response.getStatus(), equalTo(413));
             assertThat(response.getContent(), containsString("&gt;8192"));
@@ -275,7 +279,7 @@ public class SizeLimitHandlerTest
         _contextHandler.setHandler(new Handler.Abstract()
         {
             @Override
-            public boolean handle(Request request, Response response, Callback callback) throws Exception
+            public boolean handle(Request request, Response response, Callback callback)
             {
                 response.write(true, BufferUtil.toBuffer(message), callback);
                 return true;

--- a/jetty-ee10/jetty-ee10-servlet/src/main/java/org/eclipse/jetty/ee10/servlet/HttpOutput.java
+++ b/jetty-ee10/jetty-ee10-servlet/src/main/java/org/eclipse/jetty/ee10/servlet/HttpOutput.java
@@ -432,7 +432,7 @@ public class HttpOutput extends ServletOutputStream implements Runnable
         }
 
         if (content != null)
-            channelWrite(content, true, new CompleteWriteCompleteCB());
+            channelWrite(content, true, new WriteCompleteCB());
     }
 
     /**
@@ -1764,17 +1764,6 @@ public class HttpOutput extends ServletOutputStream implements Runnable
         public InvocationType getInvocationType()
         {
             return InvocationType.NON_BLOCKING;
-        }
-    }
-
-    private class CompleteWriteCompleteCB extends WriteCompleteCB
-    {
-        @Override
-        public void failed(Throwable x)
-        {
-            // TODO why is this needed for h2/h3?
-            HttpOutput.this._servletChannel.abort(x);
-            super.failed(x);
         }
     }
 }

--- a/jetty-ee10/jetty-ee10-servlet/src/main/java/org/eclipse/jetty/ee10/servlet/ServletChannelState.java
+++ b/jetty-ee10/jetty-ee10-servlet/src/main/java/org/eclipse/jetty/ee10/servlet/ServletChannelState.java
@@ -1051,6 +1051,9 @@ public class ServletChannelState
             if (_requestState != RequestState.COMPLETING)
                 throw new IllegalStateException(this.getStatusStringLocked());
 
+            if (failure != null)
+                abortResponse(failure);
+
             if (_event == null)
             {
                 _requestState = RequestState.COMPLETED;

--- a/jetty-ee10/jetty-ee10-servlet/src/test/java/org/eclipse/jetty/ee10/servlet/SizeLimitHandlerServletTest.java
+++ b/jetty-ee10/jetty-ee10-servlet/src/test/java/org/eclipse/jetty/ee10/servlet/SizeLimitHandlerServletTest.java
@@ -18,6 +18,7 @@ import java.io.ByteArrayOutputStream;
 import java.io.IOException;
 import java.net.URI;
 import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.CountDownLatch;
 import java.util.concurrent.TimeUnit;
 import java.util.zip.GZIPInputStream;
 import java.util.zip.GZIPOutputStream;
@@ -40,7 +41,6 @@ import org.eclipse.jetty.server.handler.gzip.GzipHandler;
 import org.eclipse.jetty.util.BufferUtil;
 import org.eclipse.jetty.util.IO;
 import org.junit.jupiter.api.AfterEach;
-import org.junit.jupiter.api.BeforeEach;
 import org.junit.jupiter.api.Test;
 
 import static org.hamcrest.MatcherAssert.assertThat;
@@ -49,6 +49,7 @@ import static org.hamcrest.Matchers.equalTo;
 import static org.hamcrest.Matchers.greaterThan;
 import static org.hamcrest.Matchers.lessThan;
 import static org.junit.jupiter.api.Assertions.assertNotNull;
+import static org.junit.jupiter.api.Assertions.assertTrue;
 
 public class SizeLimitHandlerServletTest
 {
@@ -57,8 +58,7 @@ public class SizeLimitHandlerServletTest
     private ServerConnector _connector;
     private HttpClient _client;
 
-    @BeforeEach
-    public void before() throws Exception
+    private void start(HttpServlet servlet) throws Exception
     {
         _server = new Server();
         _connector = new ServerConnector(_server);
@@ -71,15 +71,7 @@ public class SizeLimitHandlerServletTest
         sizeLimitHandler.setHandler(gzipHandler);
         contextHandler.insertHandler(sizeLimitHandler);
 
-        contextHandler.addServlet(new HttpServlet()
-        {
-            @Override
-            protected void doPost(HttpServletRequest req, HttpServletResponse resp) throws IOException
-            {
-                String requestContent = IO.toString(req.getInputStream());
-                resp.getWriter().print(requestContent);
-            }
-        }, "/");
+        contextHandler.addServlet(servlet, "/");
 
         _server.setHandler(contextHandler);
         _server.start();
@@ -99,6 +91,16 @@ public class SizeLimitHandlerServletTest
     @Test
     public void testGzippedEcho() throws Exception
     {
+        start(new HttpServlet()
+        {
+            @Override
+            protected void doPost(HttpServletRequest req, HttpServletResponse resp) throws IOException
+            {
+                String requestContent = IO.toString(req.getInputStream());
+                resp.getWriter().print(requestContent);
+            }
+        });
+
         String content = "x".repeat(SIZE_LIMIT * 2);
 
         URI uri = URI.create("http://localhost:" + _connector.getLocalPort());
@@ -123,6 +125,16 @@ public class SizeLimitHandlerServletTest
     @Test
     public void testNormalEcho() throws Exception
     {
+        start(new HttpServlet()
+        {
+            @Override
+            protected void doPost(HttpServletRequest req, HttpServletResponse resp) throws IOException
+            {
+                String requestContent = IO.toString(req.getInputStream());
+                resp.getWriter().print(requestContent);
+            }
+        });
+
         String content = "x".repeat(SIZE_LIMIT * 2);
 
         URI uri = URI.create("http://localhost:" + _connector.getLocalPort());
@@ -135,6 +147,18 @@ public class SizeLimitHandlerServletTest
     @Test
     public void testGzipEchoNoAcceptEncoding() throws Exception
     {
+        start(new HttpServlet()
+        {
+            @Override
+            protected void doPost(HttpServletRequest req, HttpServletResponse resp) throws IOException
+            {
+                String requestContent = IO.toString(req.getInputStream());
+                // The content will be buffered, and the implementation
+                // will try to flush it, failing because of SizeLimitHandler.
+                resp.getWriter().print(requestContent);
+            }
+        });
+
         String content = "x".repeat(SIZE_LIMIT * 2);
         URI uri = URI.create("http://localhost:" + _connector.getLocalPort());
 
@@ -150,6 +174,54 @@ public class SizeLimitHandlerServletTest
             })
             .send(resultFuture::complete);
 
+        Result result = resultFuture.get(5, TimeUnit.SECONDS);
+        assertNotNull(result);
+        assertThat(result.getResponse().getStatus(), equalTo(HttpStatus.INTERNAL_SERVER_ERROR_500));
+        assertThat(contentReceived.toString(), containsString("Response body is too large"));
+    }
+
+    @Test
+    public void testGzipEchoNoAcceptEncodingFlush() throws Exception
+    {
+        CountDownLatch flushFailureLatch = new CountDownLatch(1);
+        start(new HttpServlet()
+        {
+            @Override
+            protected void doPost(HttpServletRequest req, HttpServletResponse resp) throws IOException
+            {
+                String requestContent = IO.toString(req.getInputStream());
+                // The content will be buffered.
+                resp.getWriter().print(requestContent);
+                try
+                {
+                    // The flush will fail because exceeds
+                    // the SizeLimitHandler configuration.
+                    resp.flushBuffer();
+                }
+                catch (Throwable x)
+                {
+                    flushFailureLatch.countDown();
+                    throw x;
+                }
+            }
+        });
+
+        String content = "x".repeat(SIZE_LIMIT * 2);
+        URI uri = URI.create("http://localhost:" + _connector.getLocalPort());
+
+        StringBuilder contentReceived = new StringBuilder();
+        CompletableFuture<Result> resultFuture = new CompletableFuture<>();
+        _client.POST(uri)
+            .headers(httpFields -> httpFields.add(HttpHeader.CONTENT_ENCODING, "gzip"))
+            .body(gzipContent(content))
+            .onResponseContentAsync((response, chunk, demander) ->
+            {
+                contentReceived.append(BufferUtil.toString(chunk.getByteBuffer()));
+                demander.run();
+            })
+            .send(resultFuture::complete);
+
+        assertTrue(flushFailureLatch.await(5, TimeUnit.SECONDS));
 
         Result result = resultFuture.get(5, TimeUnit.SECONDS);
         assertNotNull(result);

